pkg/scaffold/olm-catalog/csv.go

```diff
@@ -0,0 +1,328 @@
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"unicode"
+
+	"github.com/operator-framework/operator-sdk/internal/util/yamlutil"
+	"github.com/operator-framework/operator-sdk/pkg/scaffold"
+	"github.com/operator-framework/operator-sdk/pkg/scaffold/input"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/ghodss/yaml"
+	olmApi "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	CSVYamlFileExt    = ".csv.yaml"
+	CSVConfigYamlFile = "csv-config.yaml"
+)
+
+var ErrNoCSVVersion = errors.New("no CSV version supplied")
+
+type CSV struct {
+	input.Input
+
+	// ConfigFilePath is the location of a configuration file path for this
+	// projects' CSV file.
+	ConfigFilePath string
+	// CSVVersion is the CSV current version.
+	CSVVersion string
+}
+
+func (s *CSV) GetInput() (input.Input, error) {
+	// A CSV version is required.
+	if s.CSVVersion == "" {
+		return input.Input{}, ErrNoCSVVersion
+	}
+	if s.Path == "" {
+		name := strings.ToLower(s.ProjectName) + CSVYamlFileExt
+		s.Path = filepath.Join(scaffold.OlmCatalogDir, name)
+	}
+	if s.ConfigFilePath == "" {
+		s.ConfigFilePath = filepath.Join(scaffold.OlmCatalogDir, CSVConfigYamlFile)
+	}
+	return s.Input, nil
+}
+
+// CustomRender allows a CSV to be written by marshalling
+// olmApi.ClusterServiceVersion instead of writing to a template.
+func (s *CSV) CustomRender() ([]byte, error) {
+	// Get current CSV to update.
+	csv, exists, err := getCurrentCSVIfExists(s.Path)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		csv = new(olmApi.ClusterServiceVersion)
+		s.initCSVFields(csv)
+	}
+
+	cfg, err := getCSVConfig(s.ConfigFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = s.updateCSVVersions(csv); err != nil {
+		return nil, err
+	}
+	if err = s.updateCSVFromManifestFiles(cfg, csv); err != nil {
+		return nil, err
+	}
+
+	// A new csv won't have several required fields populated.
+	if err = checkRequiredCSVFields(csv); err != nil {
+		if exists {
+			log.Warnf("Required csv fields not filled in file %s:%s\n", s.Path, err)
+		} else {
+			// Report required fields to user informationally.
+			log.Infof("Fill in the following required fields in file %s:%s\n", s.Path, err)
+		}
+	}
+
+	// Remove the status field from the CSV, as status is managed at runtime.
+	cu, err := runtime.DefaultUnstructuredConverter.ToUnstructured(csv)
+	if err != nil {
+		return nil, err
+	}
+	delete(cu, "status")
+	return yaml.Marshal(&cu)
+}
+
+func getCurrentCSVIfExists(csvPath string) (*olmApi.ClusterServiceVersion, bool, error) {
+	if _, err := os.Stat(csvPath); err != nil && os.IsNotExist(err) {
+		return nil, false, nil
+	}
+
+	csvBytes, err := ioutil.ReadFile(csvPath)
+	if err != nil {
+		return nil, false, err
+	}
+	if len(csvBytes) == 0 {
+		return nil, false, nil
+	}
+
+	csv := new(olmApi.ClusterServiceVersion)
```

Comment:
This is more of style quirk but any reason why we use `new()` (here and other places in this PR) instead of `&olmApi.ClusterServiceVersion{}`?
They're both identical but I'm wondering if we should stick to a particular convention as we do in the rest of the codebase.

Side note on why `new()` exists: https://softwareengineering.stackexchange.com/questions/210399/why-is-there-a-new-in-go/216582#216582

Reply:
The yaml library I was originally using segfaulted when using the ampersand pointer declaration style so I used `new()`. I switched to `github.com/ghodss/yaml` which does not segfault, just checked. Switching to ampersand style.